### PR TITLE
validate plotly_(api)_domain - HTTPS, not HTTP

### PR DESCRIPTION
**Pull request by Kully**
Addresses https://github.com/plotly/plotly.py/issues/996

Notes:
1. This error message does not get triggered if the original url in the request starts with `zhttp://...`. A separate error message is thrown for that.
2. The original url sent by a user is getting coerced into all lowercase, so a lowercase to lowercase comparison is not necessary.

### CODE REVIEW

**Comment by Kully** [comment]
cc @BRONSOLO @cldougl

**Comment by BRONSOLO** [comment]
Thanks @Kully! 

@Kully @cldougl what do you y'all think of approaching this fix as a validation on the `.config` file as opposed to error handling on a failed request? 

Perhaps a helper function like `validated_config_file` can be written and called in the `get/set_config_file` helper functions to make sure the structure of the domains are valid (in this case, set to use HTTPS).

**Comment by Kully** [comment]
> what do y'all think of approaching this fix as a validation on the .config file as opposed to error handling on a failed request?

I like this approach. What about we have both just in case? I don't know why the helper function could be bypassed if we set it up correctly though

**Comment by BRONSOLO** [comment]
>What about we have both just in case?

I think the validation will be sufficient, personally!

**Comment by cldougl** [comment]
Yep I think the config validation could be more beneficial for the user.
Also we suggest that on-premise users set only their `plotly_domain` rather than both `plotly_domain` and `plotly_api_domain` (https://plot.ly/python/getting-started/#special-instructions-for-plotly-onpremise-users) so I think it would be confusing to mention the `plotly_api_domain` in a message returned to the user.

**Comment by BRONSOLO** [comment]
@Kully I think this should just be issued as a warning for now and not a full error. Though SSL/TLS is now required for On-Prem, there are some clients that still may be running older versions without SSL/TLS enabled.

**Comment by Kully** [comment]
@BRONSOLO I've deleted the error msg and made a warning instead.

I only placed the `validate_config_file()` in `set_config_file()` as `get_config_file()` appears in the `set` function.

There is the possibility of added a `validate` param in `get_config` that calls the `validate_config_file` within.

I personally think that we need to validate `get_config_file` without a bunch of the same error message popping up

File: plotly/tools.py [hunk 1]
```diff
@@ -244,9 +276,10 @@ def get_config_file(*args):
     """
     if check_file_permissions():
         ensure_local_plotly_files()  # make sure what's there is OK
-        return utils.load_json_dict(CONFIG_FILE, *args)
+        returned_obj = utils.load_json_dict(CONFIG_FILE, *args)
     else:
-        return FILE_CONTENT[CONFIG_FILE]
+        returned_obj = FILE_CONTENT[CONFIG_FILE]
+    return returned_obj
```
Comment (BRONSOLO) [inline plotly/tools.py]: Is this change necessary? Could you add it to a separate commit with a commit msg that explains why if so?

File: plotly/tools.py [hunk 2]
```diff
@@ -49,6 +49,36 @@
...
+    "    plotly_api_domain='https://plotly.your-company.com'\n"
+    ")\n\n\n"
+    "Make sure to replace `your-company.com` with the URL of your Plotly "
+    "On-Premise server.\nSee "
+    "https://plot.ly/python/getting-started/#special-instructions-for-plotly-onpremise-users "
+    "for more help with getting started with On-Prem."
+)
+
+
+def validate_config_file(*domains):
```
Comment (BRONSOLO) [inline plotly/tools.py]: This is really just validating domains - what do you think about renaming this as `validate_domains`?

Comment (Kully) [reply plotly/tools.py]: good idea

**Comment by Kully** [comment]
notes:
- moved test to plotly.utils
- new test should be passing 🎉 
- only raise UserWarning on `set_config_file`, not `get_config_file`

Waiting for tests to pass.

**Comment by Kully** [comment]
Looks like only streaming errors. We are ignoring these for now @cldougl ?

**Comment by cldougl** [comment]
@kully, the streaming tests should pass now

File: plotly/tools.py [hunk 3]
```diff
@@ -186,12 +186,14 @@ def set_config_file(plotly_domain=None,
     :param (bool) world_readable: True = public, False = private
 
     """
+    # import ipdb; ipdb.set_trace()
```
Comment (BRONSOLO) [inline plotly/tools.py]: please remove this commented code as it is unnecessary 🔪 : )

File: plotly/utils.py [hunk 4]
```diff
@@ -297,6 +322,7 @@ def encode_as_decimal(obj):
         else:
             raise NotEncodable
 
+
```
Comment (BRONSOLO) [inline plotly/utils.py]: Is this line needed?

**Comment by BRONSOLO** [comment]
@Kully could you please provide a screenshot of what that warning message looks like when hit?

File: plotly/utils.py [hunk 5]
```diff
@@ -32,6 +33,30 @@
 lock = threading.Lock()
 
 
+http_msg = (
+    "The plotly_domain and plotly_api_domain of your config file must start "
+    "with 'https', 'http'.\n If you are not using On-Prem then run the "
```
Comment (BRONSOLO) [inline plotly/utils.py]: Why include `http` here? --->  "with 'https', 'http'. 

Also I would suggest using `On-Premise` instead of `On-Prem` to be consistent with the rest of the codebase.

Comment (Kully) [reply plotly/utils.py]: That was a typo. I meant to type `'http', not 'https'`

**Comment by Kully** [comment]
@BRONSOLO all comments addressed. Waiting for tests again.

**Comment by Kully** [comment]
yay tests are passing 🎉

File: plotly/utils.py [hunk 6]
```diff
@@ -35,15 +35,15 @@
 
 http_msg = (
     "The plotly_domain and plotly_api_domain of your config file must start "
-    "with 'https', 'http'.\n If you are not using On-Prem then run the "
+    "with 'https', not 'http'.\n If you are not using On-Premise then run the "
```
Comment (BRONSOLO) [inline plotly/utils.py]: I believe the `\n If` ... spacing still needs to be fixed. I think it would look cleaner without a new line actually.

**Comment by BRONSOLO** [comment]
One small comment and then 💃 after that.
